File: LeetCode/Medium/0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.java
class Solution {

    int[][] prefixCounts;
    int[] counts;
    int maxSize;

    public int characterReplacement(String s, int k) {
        prefixCounts = new int[s.length() + 1][91];
        counts = new int[91];

        for (int i = 1; i <= s.length(); i++) {
            counts[s.charAt(i - 1)]++;

            for (char c = 'A'; c <= 'Z'; c++) {
                prefixCounts[i][c] = counts[c];
            }
        }

        for (int right = s.length(); right > 0; right--) {
            maxSize = Math.max(maxSize, right - findLeft(right, k) + 1);
        }

        return maxSize;
    }

    private int findLeft(int end, int k) {
        int ret = end;
        int left = 1;
        int right = end;

        while (left <= right) {
            int mid = (left + right) >>> 1;
            int size = end - mid + 1;
            int maxCount = 0;

            if (size < maxSize) {
                right = mid - 1;
                continue;
            }

            for (char c = 'A'; c <= 'Z'; c++) {
                maxCount = Math.max(maxCount, prefixCounts[end][c] - prefixCounts[mid - 1][c]);
            }

            if (size <= maxCount + k) {
                ret = mid;
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }

        return ret;
    }

}

File: LeetCode/Medium/0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.kt
class Solution {

    fun characterReplacement(s: String, k: Int): Int {
        val counts = IntArray(26)

        var maxCount = 0
        var left = 0

        for ((right, c) in s.withIndex()) {
            maxCount = maxOf(maxCount, ++counts[c - 'A'])

            // 현재 size가 maxCount + k 보다 크다면
            // window size를 한 칸 줄이면 됨
            // 이전까지는 조건을 만족했으니까!
            if (right - left + 1 > maxCount + k) {
                counts[s[left++] - 'A']--
            }
        }

        return s.length - left
    }
}
